vital/util/interval.h

```diff
@@ -0,0 +1,139 @@
+// https://github.com/Kitware/kwiver/blob/master/LICENSE for details.
+
+/// \file
+/// \brief Definition of interval template class.
+
+#include <vital/util/vital_util_export.h>
+
+#include <algorithm>
+#include <stdexcept>
+#include <type_traits>
+
+#include <cmath>
+
+namespace kwiver {
+
+namespace vital {
+
+// ----------------------------------------------------------------------------
+/// Represents a numerical interval with an upper and lower bound.
+template < class T >
+class VITAL_UTIL_EXPORT interval
+{
+public:
+  interval( T lower, T upper )
+    : m_lower{ std::min( lower, upper ) },
+      m_upper{ std::max( lower, upper ) }
+  {
+    assert_not_nan( lower );
+    assert_not_nan( upper );
+  }
+
+  /// Return lower bound.
+  T
+  lower() const
+  {
+    return m_lower;
+  }
+
+  /// Return upper bound.
+  T
+  upper() const
+  {
+    return m_upper;
+  }
+
+  /// Ensure the lower bound is at least \p new_lower.
+  ///
+  /// \throws invalid_argument If \p new_lower is greater than the current
+  /// upper bound.
+  void
+  truncate_lower( T new_lower )
+  {
+    assert_not_nan( new_lower );
+    if( new_lower > m_upper )
+    {
+      throw std::invalid_argument(
+              "interval.truncate_lower(): "
+              "new_lower cannot be greater than current upper" );
+    }
+    m_lower = std::max( m_lower, new_lower );
+  }
+
+  /// Ensure the upper bound is at most \p new_upper.
+  ///
+  /// \throws invalid_argument If \p new_upper is less than the current lower
+  /// bound.
+  void
+  truncate_upper( T new_upper )
+  {
+    assert_not_nan( new_upper );
+    if( new_upper < m_lower )
+    {
+      throw std::invalid_argument(
+              "interval.truncate_upper(): "
+              "new_upper cannot be less than current lower" );
+    }
+    m_upper = std::min( m_upper, new_upper );
+  }
+
+  /// Ensure the interval contains \p value.
+  void
+  encompass( T value )
+  {
+    assert_not_nan( value );
+    m_lower = std::min( m_lower, value );
+    m_upper = std::max( m_upper, value );
+  }
+
+  /// Return \c true if \p value is within in this half-open interval.
+  bool
+  contains( T value ) const
+  {
+    return m_lower <= value && value < m_upper;
+  }
+
+  /// Return \c true if \p value is within in this interval.
+  ///
+  /// \param inclusive_lower Whether to return \c true if \p value is equal to
+  /// the lower bound.
+  /// \param inclusive_upper Whether to return \c true if \p value is equal to
+  /// the upper bound.
+  bool
+  contains( T value, bool inclusive_lower, bool inclusive_upper ) const
+  {
+    return ( m_lower < value || ( inclusive_lower && m_lower == value ) ) &&
+           ( value < m_upper || ( inclusive_upper && m_upper == value ) );
+  }
+
+private:
+  static void
+  assert_not_nan( T value )
+  {
+    if( std::isnan( static_cast< double >( value ) ) )
+    {
+      throw std::invalid_argument( "interval: cannot accept NaN value" );
+    }
```

Comment:
While much nicer, this is a pessimization for `float`, and *especially* for integer types.

I could live with:
```suggestion
    // TODO(C++17): make if constexpr and remove the static_cast
    if( std::is_floating_point< T >::value )
    {
      using promoted_t = std::common_type<T, float>::type;
      if( std::isnan( static_cast< promoted_t >( value ) ) )
      {
        throw std::invalid_argument( "interval: cannot accept NaN value" );
      }
    }
```

...but I'd be happier if MSVC's standard library was conforming 😢. (See also microsoft/STL#519.)

That said, GCC and Clang are smart enough to optimize away the test in spite of the cast (but still add an instruction to the `float` case; hence the `promoted_t`), so if you wanted to just give MSVC the finger (performance-wise), you could omit the `is_floating_point` check.

Reply:
At this point, should I just revert back to the original, `enable_if`-based version? No performance hit, never needs to change, and not too much harder to understand than this... Or perhaps add a `isnan_integral_friendly()` (name WIP) function in a utility file somewhere, with overloads for integral types? I anticipate needing something like `assert_not_nan()` again in another class.

Reply:
> the original [...] never needs to change

I would disagree with *that* 🙂. Even my suggestion is *much* cleaner than the original version, and I believe is equally (and optimally) performant.

OTOH, if you expect to need this a bunch going forward, what I would suggest is (partially pseudocode):
```c++
namespace vital {
#if USING_BROKEN_MSVC (replace with proper version check)
// See https://github.com/microsoft/STL/issues/519
template < typename T > bool isnan( T value )
{
  if constexpr ( std::is_floating_point_v< T > )
  {
    return std::isnan( value );
  }
  else
  {
    return false;
  }
}
#else
using std::isnan;
#endif
}
```

If we can't rely on C++17 MSVC, you could use the code from my previous comment instead. Major point, though, is I would just call it `vital::isnan` 🙂.